internal/provider/client.go

```diff
@@ -0,0 +1,454 @@
+package provider
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"math"
+	"math/rand"
+	"time"
+
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+
+	"github.com/Jeffail/gabs/v2"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"golang.org/x/net/html"
+)
+
+const ndAuthPayload = `{
+	"userName": "%s",
+	"userPasswd": "%s"
+}`
+
+// Default timeout for NGINX in ND is 90 Seconds.
+// Allow the client to set a shorter or longer time depending on their
+// environment
+const DefaultReqTimeoutVal int = 100
+const DefaultBackoffMinDelay int = 4
+const DefaultBackoffMaxDelay int = 60
+const DefaultBackoffDelayFactor float64 = 3
+
+// Client is the main entry point
+type Client struct {
+	BaseURL            *url.URL
+	httpClient         *http.Client
+	AuthToken          *Auth
+	Mutex              sync.Mutex
```

Comment:
Do we need BaseURL, AuthToken and Mutex to be externally accessible? should we make them start with a lower case character?

Reply:
@sajagana can you please elaborate on if we need BaseURL, AuthToken and Mutex to be externally accessible?

Reply:
@anvitha-jain Currently, we do not access the BaseURL, Mutex, and AuthToken outside the client. We can change it if required.